pydantic/fields.py

```diff
@@ -439,16 +439,43 @@ def merge_field_infos(*field_infos: FieldInfo, **overrides: Any) -> FieldInfo:
 
         new_kwargs: dict[str, Any] = {}
         metadata = {}
+        schema_callables, schema_dict = set(), {}
         for field_info in field_infos:
+            json_schema_extra: JsonDict | typing.Callable[[JsonDict], None] | None = field_info._attributes_set.get(
+                'json_schema_extra'
+            )  # type: ignore
+            if json_schema_extra:
+                if callable(json_schema_extra):
+                    schema_callables.add(json_schema_extra)
+                else:
+                    schema_dict.update(json_schema_extra)
```

Comment:
So my approach is that, for mixed type, will gather all the `json_schema_extra` first. Then push them in a `callable` so it will behave like normal `callable json_schema_extra`. In there, the `dict`s will merge into a final single `dict`, the `callable` will got called as usual.

If not then treat them as normal.

Reply:
I'm wondering why the behavior is currently different for type adapters vs model fields? Can we reference the current logic for annotated types to see what might be the best approach here?

I'm not entirely opposed to this, just curious.

Reply:
Sure. @sydney-runkle 
The current logic for model fields is that it go merging the Fields by update the entire attrs `new_kwargs.update(field_info._attributes_set)` so only the last `json_schema_extra` will be available. Then it go eval that json_schema_extra.

Type adapters just simply eval each `json_schema_extra`.

Reply:
@nix010,

Where are you seeing that logic for `TypeAdapter`?

Reply:
@sydney-runkle 
```python
        # _internal/_generate_schema.py: 1820
        for annotation in annotations:
            if annotation is None:
                continue
            get_inner_schema = self._get_wrapped_inner_schema(
                get_inner_schema, annotation, pydantic_js_annotation_functions
```
This piece of code will generate the schema with wrapper func for each of the `extra_json_schema` in each `Field`.
<img width="1592" alt="image" src="https://github.com/pydantic/pydantic/assets/16438204/2370e25d-1ba3-4076-8977-97f684493fc9">
and the `.json_schema()` will eval each of them after.

Reply:
@nix010,

Great find. Will take a closer look at the technical internals of your work here this afternoon!